### PR TITLE
Use Unix F77 convention for logical representation

### PR DESCRIPTION
**Pull request by jeanPerier**
Follow UNIX F77 convention for logicals:
- underlying integer is not zero => logical is .TRUE.
- logical is .TRUE. => set underlying integer to 1.

This matches gfortran and nvfortran -Munixlogical behavior, and the one from the front-end type representation.

(The previous behavior was taken from integer conversion and had .true. being represented -1, and lowest bit being one being evaluated as true).

### CODE REVIEW

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 1]
```diff
@@ -861,6 +862,25 @@ struct ConvertOpConversion : public FIROpConversion<fir::ConvertOp> {
...
       return mlir::success();
     }
+
+    // Follow UNIX F77 convention for logicals:
+    // 1. underlying integer is not zero => logical is .TRUE.
+    // 2. logical is .TRUE. => set underlying integer to 1.
+    auto isI1Type = [](mlir::Type ty) {
+      return ty.isa<mlir::IntegerType>() &&
+             mlir::LLVM::getPrimitiveTypeSizeInBits(ty) == 1;
+    };
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:872]: There is exactly one instance of `i1` type in the MLIRContext, so a simple comparison is all that is needed.

```c++
  i1Ty = mlir::IntegerType::get(ctx, 1);
  if (typeOfThing == i1Ty) ...
```

**Review by schweitzpgi — COMMENTED** [review]
Thanks, Jean.

File: flang/lib/Lower/IO.cpp [hunk 2]
```diff
@@ -526,8 +542,12 @@ static mlir::Value createIoRuntimeCallForItem(mlir::Location loc,
                    itemTy.cast<mlir::IntegerType>().getWidth() / 8)));
     }
   }
-  return builder.create<fir::CallOp>(loc, inputFunc, inputFuncArgs)
-      .getResult(0);
+  auto call = builder.create<fir::CallOp>(loc, inputFunc, inputFuncArgs);
+  auto itemAddr = fir::getBase(item);
+  auto itemTy = fir::unwrapPassByRefType(itemAddr.getType());
+  if (itemTy.isa<fir::LogicalType>())
+    boolRefToLogical(loc, builder, itemAddr);
```
Comment (schweitzpgi) [inline flang/lib/Lower/IO.cpp:549]: Seems like another approach might be to normalize the LOGICAL in memory to be `0` or `1` before the call. It sounds like the runtime doesn't use/know about the more significant bytes.

File: flang/lib/Lower/IO.cpp [hunk 3]
```diff
@@ -500,6 +500,22 @@ static mlir::FuncOp getInputFunc(mlir::Location loc, fir::FirOpBuilder &builder,
   return getIORuntimeFunc<mkIOKey(InputDescriptor)>(loc, builder);
 }
 
+/// Interpret addr as a C++ bool, and then set its value to the equivalent
+/// Fortran logical. This needs to be called for logical that were passed as
+/// bool& to the runtime and whose upper bytes, if any, may not be correct
+/// anymore.
```
Comment (schweitzpgi) [inline flang/lib/Lower/IO.cpp:506]: Maybe mention that C++ bool == a byte?

**Comment by jeanPerier** [comment]
Outdated, some regressions to be investigated first.

**Pull request closed without merging**